File: algorithms/code/leetcode/lc120_triangle/LC120Triangle.java
package leetcode.lc120_triangle;

import java.util.List;

/**
 * https://leetcode.com/problems/triangle/
 * #medium
 */
public final class LC120Triangle {
    public int minimumTotal(final List<List<Integer>> triangle) {
        if (triangle == null || triangle.size() == 0 || triangle.get(0).size() == 0) {
            return 0;
        }
        int height = triangle.size();
        int[] sums = new int[triangle.get(height - 1).size()];
        sums[0] = triangle.get(0).get(0);
        for (int l = 1; l < height; l++) {
            List<Integer> line = triangle.get(l);
            int n = line.size();
            for (int i = n - 1; i >= 0; i--) {
                int left = i == 0 ? Integer.MAX_VALUE : sums[i - 1];
                int right = i == n - 1 ? Integer.MAX_VALUE : sums[i];
                sums[i] = Math.min(left, right) + line.get(i);
            }
        }
        int min = sums[0];
        for (int i = 1; i < sums.length; i++) {
            if (sums[i] < min) {
                min = sums[i];
            }
        }
        return min;
    }
}
package leetcode.lc120_triangle;

import java.util.ArrayList;
import java.util.List;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public final class LC120TriangleTests {
    private List<List<Integer>> construct(final int[][] compact) {
        List<List<Integer>> triangle = new ArrayList<>();
        for (int[] array : compact) {
            List<Integer> line = new ArrayList<>();
            triangle.add(line);
            for (int value : array) {
                line.add(value);
            }
        }
        return triangle;
    }

    @Test
    public void testExample() throws Exception {
        int[][] triangle = {
                { 2 },
                { 3, 4 },
                { 6, 5, 7 },
                { 4, 1, 8, 3 }
        };
        assertEquals(11, new LC120Triangle().minimumTotal(construct(triangle)));
    }

    @Test
    public void testNothing() throws Exception {
        assertEquals(0, new LC120Triangle().minimumTotal(null));
        assertEquals(0, new LC120Triangle().minimumTotal(new ArrayList<>()));
        List<List<Integer>> nothing = new ArrayList<>();
        nothing.add(new ArrayList<>());
        assertEquals(0, new LC120Triangle().minimumTotal(nothing));
    }
}

File: algorithms/code/leetcode/lc120_triangle/lc120_triangle.py
#!/usr/bin/env python3
# https://leetcode.com/problems/triangle/

import unittest
from typing import List


class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        if triangle is None or len(triangle) == 0 or len(triangle[0]) == 0:
            return 0
        height = len(triangle)
        sums = [0] * len(triangle[height - 1])
        sums[0] = triangle[0][0]
        for i in range(1, height):
            line = triangle[i]
            n = len(line)
            for j in range(n - 1, -1, -1):
                left = float('inf') if j == 0 else sums[j - 1]
                right = float('inf') if j == n - 1 else sums[j]
                sums[j] = min(left, right) + line[j]
        minimum = sums[0]
        for i, _ in enumerate(sums):
            if sums[i] < minimum:
                minimum = sums[i]
        return minimum


class TestCode(unittest.TestCase):
    def __construct(self, compact):
        triangle = []
        for array in compact:
            line = []
            triangle.append(line)
            for value in array:
                line.append(value)
        return triangle

    def test_example(self):
        triangle = [[2], [3, 4], [6, 5, 7], [4, 1, 8, 3]]
        self.assertEqual(
            11, Solution().minimumTotal(self.__construct(triangle))
        )

    def test_nothing(self):
        self.assertEqual(0, Solution().minimumTotal(None))
        self.assertEqual(0, Solution().minimumTotal([]))
        self.assertEqual(0, Solution().minimumTotal([[]]))
